File: android-compose/app/src/main/java/com/ohmz/tday/compose/core/navigation/AppRoute.kt
package com.ohmz.tday.compose.core.navigation

import android.net.Uri

sealed class AppRoute(val route: String) {
    data object Splash : AppRoute("splash")
    data object ServerSetup : AppRoute("server-setup")
    data object Login : AppRoute("login")
    data object ForgotPassword : AppRoute("forgot-password")
    data object ScheduledTaskHome : AppRoute("home")
    data object FloaterTaskHome : AppRoute("floater")
    data object TodayTodos : AppRoute("todos/today")
    data object CreateTodayTodo : AppRoute("todos/create?target={target}")
    data object OverdueTodos : AppRoute("todos/overdue")
    data object ScheduledTodos : AppRoute("todos/scheduled")
    data object AllTodos : AppRoute("todos/all?highlightTodoId={highlightTodoId}") {
        fun create(highlightTodoId: String? = null): String {
            return if (highlightTodoId.isNullOrBlank()) {
                "todos/all"
            } else {
                "todos/all?highlightTodoId=${Uri.encode(highlightTodoId)}"
            }
        }
    }
    data object PriorityTodos : AppRoute("todos/priority")
    data object ListTodos : AppRoute("todos/list/{listId}/{listName}") {
        fun create(listId: String, listName: String): String {
            return "todos/list/$listId/${Uri.encode(listName)}"
        }
    }
    data object FloaterListTodos : AppRoute("floater/list/{listId}/{listName}") {
        fun create(listId: String, listName: String): String {
            return "floater/list/$listId/${Uri.encode(listName)}"
        }
    }

    data object Completed : AppRoute("completed?scope={scope}") {
        /**
         * The route with no tab named — read off the pattern rather than spelled
         * a second time, so the two cannot drift.
         */
        private val unscoped: String = route.substringBefore('?')

        /**
         * The completion history, optionally opened on one of its two tabs.
         *
         * The scope rides as a query parameter for the reason `AllTodos`'s
         * highlight does: the argument is optional, its `navArgument` declares a
         * null default, so the bare `"completed"` still matches this pattern and
         * every arrival that names no tab — the deep link, a shortcut, a
         * notification — opens the first one.
         */
        fun create(scope: CompletedScope? = null): String =
            if (scope == CompletedScope.Floater) "$unscoped?scope=${CompletedScope.Floater.wire}" else unscoped
    }
    data object Calendar : AppRoute("calendar")
    data object Car : AppRoute("car")
    data object Settings : AppRoute("settings")
    data object LatestRelease : AppRoute("latest-release")
    data object MorningSweep : AppRoute("morning-sweep")
    data object HelpGuide : AppRoute("help-guide?topic={topic}") {
        fun create(topic: String? = null): String {
            return if (topic.isNullOrBlank()) "help-guide" else "help-guide?topic=${Uri.encode(topic)}"
        }
    }
}

File: ios-swiftUI/Tday/Core/Navigation/AppRoute.swift
import Foundation

enum AppRoute: Hashable {
    case home
    case todayTodos
    case createTodayTodo
    case createFloaterTodo
    case overdueTodos
    case scheduledTodos
    case allTodos(highlightTodoId: String?)
    case priorityTodos
    case floaterTodos
    case floaterListTodos(listId: String, listName: String)
    case listTodos(listId: String, listName: String)
    case completed
    case calendar
    case settings
    case latestRelease
    case forgotPassword

    var deepLinkPath: String {
        switch self {
        case .home:
            return "home"
        case .todayTodos:
            return "todos/today"
        case .createTodayTodo:
            return "todos/create?target=today"
        case .createFloaterTodo:
            return "todos/create?target=floater"
        case .overdueTodos:
            return "todos/overdue"
        case .scheduledTodos:
            return "todos/scheduled"
        case let .allTodos(highlightTodoId):
            if let highlightTodoId, !highlightTodoId.isEmpty {
                return "todos/all?highlightTodoId=\(highlightTodoId)"
            }
            return "todos/all"
        case .priorityTodos:
            return "todos/priority"
        case .floaterTodos:
            return "floater"
        case let .floaterListTodos(listId, listName):
            return "floater/list/\(listId)/\(listName)"
        case let .listTodos(listId, listName):
            return "todos/list/\(listId)/\(listName)"
        case .completed:
            return "completed"
        case .calendar:
            return "calendar"
        case .settings:
            return "settings"
        case .latestRelease:
            return "latest-release"
        case .forgotPassword:
            return "forgot-password"
        }
    }

    static func from(url: URL) -> AppRoute? {
        guard url.scheme?.lowercased() == "tday" else {
            return nil
        }

        let host = url.host?.lowercased()
        let pathComponents = url.pathComponents.filter { $0 != "/" }
        let components = ([host].compactMap { $0 } + pathComponents)

        guard let first = components.first else {
            return .home
        }

        switch first {
        case "home":
            return .home
        case "completed":
            return .completed
        case "calendar":
            return .calendar
        case "settings":
            return .settings
        case "latest-release":
            return .latestRelease
        case "forgot-password":
            return .forgotPassword
        case "todos":
            let second = components.dropFirst().first ?? ""
            switch second {
            case "today":
                return .todayTodos
            case "create":
                let target = URLComponents(url: url, resolvingAgainstBaseURL: false)?
                    .queryItems?
                    .first(where: { $0.name == "target" })?
                    .value
                switch target {
                case "today":
                    return .createTodayTodo
                case "floater":
                    return .createFloaterTodo
                default:
                    return nil
                }
            case "overdue":
                return .overdueTodos
            case "scheduled":
                return .scheduledTodos
            case "all":
                let highlightTodoId = URLComponents(url: url, resolvingAgainstBaseURL: false)?
                    .queryItems?
                    .first(where: { $0.name == "highlightTodoId" })?
                    .value
                return .allTodos(highlightTodoId: highlightTodoId)
            case "priority":
                return .priorityTodos
            case "floater":
                let remaining = Array(components.dropFirst(2))
                if remaining.first == "list", remaining.count >= 3 {
                    return .floaterListTodos(listId: remaining[1], listName: remaining[2])
                }
                return .floaterTodos
            case "list":
                let remaining = Array(components.dropFirst(2))
                guard remaining.count >= 2 else {
                    return nil
                }
                return .listTodos(listId: remaining[0], listName: remaining[1])
            default:
                return nil
            }
        default:
            if first == "floater" {
                let remaining = Array(components.dropFirst())
                if remaining.first == "list", remaining.count >= 3 {
                    return .floaterListTodos(listId: remaining[1], listName: remaining[2])
                }
                return .floaterTodos
            }
            return nil
        }
    }
}
